kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+                class_weight_gap_tags = get_suggestion(
+                    trial,
+                    'class_weight_gap_tags',
+                    self.config.options.class_weights.gap_tags,
+                )
+                base_config_dict['system']['model']['outputs']['word_level'][
+                    'class_weights'
+                ]['gap_tags'] = {const.BAD: class_weight_gap_tags}
+                search_values['class_weight_gap_tags'] = class_weight_gap_tags
+            if (
+                self.config.options.class_weights.source_tags
+                and word_level_config['source']
+            ):
+                class_weight_source_tags = get_suggestion(
+                    trial,
+                    'class_weight_source_tags',
+                    self.config.options.class_weights.source_tags,
+                )
+                base_config_dict['system']['model']['outputs']['word_level'][
+                    'class_weights'
+                ]['source_tags'] = {const.BAD: class_weight_source_tags}
+                search_values['class_weight_source_tags'] = class_weight_source_tags
+        return train.Configuration(**base_config_dict), search_values
+
+    def __call__(self, trial) -> float:
+        # TODO: integrate PTL trainer callback
+        # ptl_callback = PyTorchLightningPruningCallback(trial, monitor='val_PEARSON')
+        # train_info = train.run(trainconfig, ptl_callback=ptl_callback)
+
+        train_config, search_values = self.suggest_train_config(trial)
+
+        logger.info(f'############# STARTING TRIAL {trial.number} #############')
+        logger.info(f'PARAMETERS: {search_values}')
+        for name, value in search_values.items():
+            logger.info(f'{name}: {value}')
+
+        try:
+            train_info = train.run(train_config)
+        except RuntimeError as e:
+            logger.info(f'ERROR OCCURED; SKIPPING TRIAL: {e}')
+            return -1
+
+        logger.info(f'############# TRIAL {trial.number} FINISHED #############')
+        result = train_info.best_metrics.get(self.main_metric, -1)
+        if result == -1:
+            logger.error(
+                'Trial did not converge:\n'
+                f'    `train_info.best_metrics={train_info.best_metrics}`\n'
+                f'Setting result to -1'
+            )
+
+        logger.info(f'RESULTS: {result} {self.main_metric}')
+        logger.info(f'MODEL: {train_info.best_model_path}')
+        logger.info(f'PARAMETERS: {search_values}')
+        for name, value in search_values.items():
+            logger.info(f'{name}: {value}')
+
+        # Store the checkpoint path for later access
+        self.model_paths.append((train_info.best_model_path, result))
+
+        return result
+
+
+def setup_run(directory: Path, seed: int, debug=False, quiet=False) -> Path:
+    # TODO: replace all of this with the MLFlow integration: optuna.integration.mlflow
```

Comment:
I see MLFlow integration on line 584 so what is missing here?

Reply:
TBH I am not sure how the integration is working now 😬